agent/hcp/client/metrics_client.go

```diff
@@ -0,0 +1,150 @@
+type cloudConfig interface {
+	HCPConfig(opts ...hcpcfg.HCPConfigOption) (hcpcfg.HCPConfig, error)
+}
+
+// otlpClient is an implementation of MetricsClient with a retryable http client for retries and to honor throttle.
+// It also holds default HTTP headers to add to export requests.
+type otlpClient struct {
+	client *retryablehttp.Client
+	header *http.Header
+}
+
+// NewMetricsClient returns a configured MetricsClient.
+// The current implementation uses otlpClient to provide retry functionality.
+func NewMetricsClient(cfg cloudConfig, logger hclog.Logger) (MetricsClient, error) {
+	if cfg == nil {
+		return nil, fmt.Errorf("failed to init telemetry client: provide valid cloudCfg (Cloud Configuration for TLS)")
+	}
+
+	if logger == nil {
+		return nil, fmt.Errorf("failed to init telemetry client: provide a valid logger")
+	}
+
+	c, err := newHTTPClient(cfg, logger)
+	if err != nil {
+		return nil, fmt.Errorf("failed to init telemetry client: %v", err)
+	}
+
+	header := make(http.Header)
+	header.Set("Content-Type", "application/x-protobuf")
+
+	return &otlpClient{
+		client: c,
+		header: &header,
+	}, nil
+}
+
+// newHTTPClient configures the retryable HTTP client.
+func newHTTPClient(cloudCfg cloudConfig, logger hclog.Logger) (*retryablehttp.Client, error) {
+	hcpCfg, err := cloudCfg.HCPConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	tlsTransport := cleanhttp.DefaultPooledTransport()
+	tlsTransport.TLSClientConfig = hcpCfg.APITLSConfig()
+
+	var transport http.RoundTripper = &oauth2.Transport{
+		Base:   tlsTransport,
+		Source: hcpCfg,
+	}
+
+	client := &http.Client{
+		Transport: transport,
+		Timeout:   defaultStreamTimeout,
+	}
+
+	retryClient := &retryablehttp.Client{
+		HTTPClient:   client,
+		Logger:       logger.Named("hcp_telemetry_client"),
+		RetryWaitMin: defaultRetryWaitMin,
+		RetryWaitMax: defaultRetryWaitMax,
+		RetryMax:     defaultRetryMax,
+		CheckRetry:   retryablehttp.DefaultRetryPolicy,
+		Backoff:      retryablehttp.DefaultBackoff,
+	}
+
+	return retryClient, nil
+}
+
+// ExportMetrics is the single method exposed by MetricsClient to export OTLP metrics to the desired HCP endpoint.
+// The endpoint is configurable as the endpoint can change during periodic refresh of CCM telemetry config.
+// By configuring the endpoint here, we can re-use the same client and override the endpoint when making a request.
+func (o *otlpClient) ExportMetrics(ctx context.Context, protoMetrics *metricpb.ResourceMetrics, endpoint string) error {
+	pbRequest := &colmetricpb.ExportMetricsServiceRequest{
+		ResourceMetrics: []*metricpb.ResourceMetrics{protoMetrics},
+	}
+
+	body, err := proto.Marshal(pbRequest)
+	if err != nil {
+		return fmt.Errorf("failed to export metrics: %v", err)
+	}
+
+	req, err := retryablehttp.NewRequest(http.MethodPost, endpoint, bytes.NewBuffer(body))
+	if err != nil {
+		return fmt.Errorf("failed to export metrics: %v", err)
+	}
+	req.Header = (*o.header).Clone()
```

Comment:
@loshz @chapmanc I think it's best to clone this. Usually, clients will hold a copy of the request and copy the whole request.